File: Linked List/2181. Merge Nodes in Between Zeros/merge_nodes.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeNodes(head *ListNode) *ListNode {
	dummyHead := &ListNode{Val: 0} // 直接初始化dummyHead
	current := dummyHead           // current指向dummyHead
	sum := 0

	for head != nil {
		if head.Val > 0 {
			sum += head.Val
		} else if sum != 0 {
			newNode := &ListNode{Val: sum} // 创建新节点
			current.Next = newNode         // 连接新节点
			current = current.Next         // 移动current到新节点
			sum = 0
		}
		head = head.Next
	}

	return dummyHead.Next
}

func mergeNodes1(head *ListNode) *ListNode {
	dummyHead := &ListNode{} // 创建一个哑节点作为头节点
	current := dummyHead     // current指向哑节点
	sum := 0                 // 初始化sum为0

	for head != nil {
		if head.Val == 0 && sum != 0 { // 如果遇到0且sum不为0，说明一个区间结束
			current.Next = &ListNode{Val: sum} // 创建新节点，值为sum，连接到current
			current = current.Next             // 更新current指向新节点
			sum = 0                            // 重置sum为0
		}
		if head.Val != 0 { // 如果不是0，累加到sum
			sum += head.Val
		}
		head = head.Next // 移动到下一个节点
	}

	return dummyHead.Next
}

// helper函数用于打印链表
func printList(head *ListNode) {
	for head != nil {
		fmt.Printf("%d ", head.Val)
		head = head.Next
	}
	fmt.Println()
}

func main() {
	// 创建链表 [0,3,1,0,4,5,2,0]
	head := &ListNode{Val: 0}
	current := head
	values := []int{3, 1, 0, 4, 5, 2, 0}
	for _, val := range values {
		current.Next = &ListNode{Val: val}
		current = current.Next
	}

	// 调用mergeNodes函数
	mergedHead := mergeNodes(head.Next)
	printList(mergedHead)

	// 创建链表 [0,1,0,3,0,2,2,0]
	head1 := &ListNode{Val: 0}
	current1 := head1
	values1 := []int{1, 0, 3, 0, 2, 2, 0}
	for _, val := range values1 {
		current1.Next = &ListNode{Val: val}
		current1 = current1.Next
	}

	// 调用mergeNodes1函数
	mergedHead1 := mergeNodes1(head1.Next)
	printList(mergedHead1)
}

File: Linked List/2181. Merge Nodes in Between Zeros/merge_nodes.rs
// 定义单链表结构体
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    // 构造函数
    #[inline]
    pub fn new(val: i32) -> Self {
        ListNode { 
            next: None, 
            val 
        }
    }
}

pub fn merge_nodes(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut current_head = head; // 创建一个新的可变变量来持有head的值
    let mut dummy_head = Box::new(ListNode::new(0));
    let mut current = &mut dummy_head;
    let mut sum = 0;

    while let Some(node) = current_head {
        if node.val > 0 {
            sum += node.val;
        } else if sum != 0 {
            current.next = Some(Box::new(ListNode::new(sum)));
            current = current.next.as_mut().unwrap();
            sum = 0;
        }
        current_head = node.next; // 修改的是current_head而不是head
    }

    dummy_head.next
}

fn main() {
    // 创建链表 [0,1,0,3,0,2,2,0]
    let mut head = Some(Box::new(ListNode::new(0)));
    let mut current = head.as_mut().unwrap();
    
    let values = vec![1, 0, 3, 0, 2, 2, 0];
    for &val in values.iter() {
        current.next = Some(Box::new(ListNode::new(val)));
        current = current.next.as_mut().unwrap();
    }

    // 调用merge_nodes函数
    let merged_head = merge_nodes(head);

    // 打印合并后的链表
    let mut current = merged_head;
    while let Some(node) = current {
        print!("{} ", node.val);
        current = node.next;
    }
    println!("");
}